File: internal/authz/approval_policy.go
// Fiscal convention: monetary values in the Drenyra ecosystem are int64 cents;
// no float is ever used for money. This module is the versioned, PURE approval
// authorization policy (v0.4.0 Step 1, ADR-003). It works from the
// PRE-VERIFIED principal only — no database, clock, token, or identity-provider
// access. The engine records the claim for provenance; this policy authorizes.
//
// Frozen v0.4.0 matrix:
//   - Base role by fiscal effect: journal_entry/adjustment/reclassification →
//     accountant; approval → senior_accountant; closing → controller;
//     declaration → tax_reviewer; sunat_filing → authorized_tax_professional.
//     Any other (or missing) effect cannot be approved under policy.
//   - Materiality raises the accounting ladder: material → senior_accountant,
//     critical → controller (only for effects whose base role is accountant or
//     senior_accountant; normal/NULL stays at base).
//   - Role dominance applies ONLY within the accounting ladder
//     accountant < senior_accountant < controller; tax roles are explicit and
//     never implied by a controller (and vice versa).
//   - Minimum assurance is standard; sunat_filing additionally requires strong.
//   - Check order (first frozen code wins, for reproducibility): tenant →
//     company scope → membership active → role → assurance → materiality.
package authz

import (
	"github.com/arkelythex/drenyra-engram/internal/auth"
	"github.com/arkelythex/drenyra-engram/internal/core"
)

// PolicyVersion is the frozen version of the policy. It is stamped on every
// decision and persisted with approval events.
const PolicyVersion = "approval-policy/v0.4.0"

// ReasonAuthorized is the reason code of an allowed decision (the approval
// event schema records authorization_reason_code='AUTHORIZED').
const ReasonAuthorized = "AUTHORIZED"

// Decision is the pure authorization outcome: allowed, the exact policy
// version, and the frozen reason code.
type Decision struct {
	Allowed      bool
	PolicyVersion string
	ReasonCode   string
}

// ApprovalAuthorizationPolicy is the pure authorization contract. Implementors
// must not touch the database, the clock, tokens, or identity providers.
type ApprovalAuthorizationPolicy interface {
	Authorize(principal auth.VerifiedApprovalPrincipal, memory core.AccountingMemory) Decision
}

// frozenPolicy is the immutable v0.4.0 policy.
type frozenPolicy struct{}

// NewApprovalPolicy returns the frozen v0.4.0 policy.
func NewApprovalPolicy() ApprovalAuthorizationPolicy { return frozenPolicy{} }

// Authorize evaluates the pre-verified principal against the memory's fiscal
// effect, declared materiality level and scope. It returns the FIRST frozen
// reason code in the fixed check order so denials are reproducible.
func (frozenPolicy) Authorize(p auth.VerifiedApprovalPrincipal, m core.AccountingMemory) Decision {
	// 1. Tenant scope: the principal's tenant must equal the memory's tenant.
	if p.TenantID() != m.Scope.OrganizationID {
		return denied(auth.CodeTenantScopeMismatch)
	}
	// 2. Company scope: the memory's company must be inside the membership's
	//    company scopes. Institutional memories have no company to authorize.
	if m.Scope.Kind != core.ScopeKindCompany || !contains(p.CompanyScopes(), m.Scope.CompanyID) {
		return denied(auth.CodeCompanyScopeDenied)
	}
	// 3. Membership active: a pre-verified principal always carries the
	//    membership id of the ACTIVE membership it was derived from; empty
	//    means no active membership (the domain does not query the identity
	//    provider — this policy works from the pre-verified principal).
	if p.MembershipID() == "" {
		return denied(auth.CodeMembershipInactive)
	}
	// 4. Role: the principal must satisfy the BASE role for the effect.
	base, ok := baseRole(m.FiscalEffect)
	if !ok || !satisfies(p.Roles(), base) {
		return denied(auth.CodeRoleNotAuthorized)
	}
	// 5. Assurance: minimum standard; sunat_filing additionally requires strong.
	minAssurance := auth.AssuranceStandard
	if m.FiscalEffect == core.FiscalEffectSunatFiling {
		minAssurance = auth.AssuranceStrong
	}
	if assuranceRank(p.AssuranceLevel()) < assuranceRank(minAssurance) {
		return denied(auth.CodeAssuranceTooLow)
	}
	// 6. Materiality: the declared level may raise the requirement above the
	//    principal's reach → MATERIALITY_LIMIT_EXCEEDED.
	raised := requiredRole(base, m.MaterialityLevel)
	if raised != base && !satisfies(p.Roles(), raised) {
		return denied(auth.CodeMaterialityLimitExceeded)
	}
	return Decision{Allowed: true, PolicyVersion: PolicyVersion, ReasonCode: ReasonAuthorized}
}

func denied(code string) Decision {
	return Decision{Allowed: false, PolicyVersion: PolicyVersion, ReasonCode: code}
}

// baseRole maps a fiscal effect to its base accounting role. Effects without a
// base role (none, or unknown) return ok=false — a memory without a fiscal
// effect cannot be approved under policy.
func baseRole(e core.FiscalEffect) (auth.AccountingRole, bool) {
	switch e {
	case core.FiscalEffectJournalEntry, core.FiscalEffectAdjustment, core.FiscalEffectReclassification:
		return auth.RoleAccountant, true
	case core.FiscalEffectApproval:
		return auth.RoleSeniorAccountant, true
	case core.FiscalEffectClosing:
		return auth.RoleController, true
	case core.FiscalEffectDeclaration:
		return auth.RoleTaxReviewer, true
	case core.FiscalEffectSunatFiling:
		return auth.RoleAuthorizedTaxProfessional, true
	default:
		return "", false
	}
}

// requiredRole raises the base role for the declared materiality level. The
// raise applies ONLY to effects whose base role is accountant or
// senior_accountant; normal/NULL stays at base; tax bases never raise.
func requiredRole(base auth.AccountingRole, level *core.MaterialityLevel) auth.AccountingRole {
	if level != nil {
		switch *level {
		case core.MaterialityCritical:
			if base == auth.RoleAccountant || base == auth.RoleSeniorAccountant {
				return auth.RoleController
			}
		case core.MaterialityMaterial:
			if base == auth.RoleAccountant {
				return auth.RoleSeniorAccountant
			}
		}
	}
	return base
}

// satisfies reports whether any principal role dominates the required role.
// Dominance exists ONLY within the accounting ladder
// (accountant < senior_accountant < controller); tax roles are matched
// explicitly — a controller NEVER implies tax_reviewer or
// authorized_tax_professional, and vice versa.
func satisfies(roles []auth.AccountingRole, required auth.AccountingRole) bool {
	if required == auth.RoleTaxReviewer || required == auth.RoleAuthorizedTaxProfessional {
		for _, r := range roles {
			if r == required {
				return true
			}
		}
		return false
	}
	rr := ladderRank(required)
	for _, r := range roles {
		if ladderRank(r) >= rr {
			return true
		}
	}
	return false
}

// ladderRank maps a role to its accounting-ladder rank. Tax roles (and any
// unknown role) sit OUTSIDE the ladder at rank 0 and can never dominate.
func ladderRank(r auth.AccountingRole) int {
	switch r {
	case auth.RoleAccountant:
		return 1
	case auth.RoleSeniorAccountant:
		return 2
	case auth.RoleController:
		return 3
	default:
		return 0
	}
}

// assuranceRank orders assurance levels: low < standard < strong. Unknown
// levels rank below everything and fail the minimum-assurance check.
func assuranceRank(l auth.AssuranceLevel) int {
	switch l {
	case auth.AssuranceLow:
		return 0
	case auth.AssuranceStandard:
		return 1
	case auth.AssuranceStrong:
		return 2
	default:
		return -1
	}
}

func contains(scopes []string, target string) bool {
	for _, s := range scopes {
		if s == target {
			return true
		}
	}
	return false
}

File: authz/approval-policy.ts
/**
 * Fiscal convention: monetary values in the Drenyra ecosystem are BigInt cents;
 * no float is ever used for money.
 *
 * Versioned approval authorization policy (v0.4.0 Step 1, ADR-003) — the exact
 * TypeScript mirror of `internal/authz/approval_policy.go`. PURE: no database,
 * clock, token, or identity-provider access; it authorizes from the
 * pre-verified principal only.
 *
 * Frozen v0.4.0 matrix and check order (first reason code wins, for
 * reproducibility): tenant → company scope → membership active → role →
 * assurance → materiality.
 */

import type {
	AccountingMemory,
	AccountingRole,
	ApprovalAuthorizationDecision,
	ApprovalErrorCode,
	AssuranceLevel,
	FiscalEffect,
	MaterialityLevel,
	VerifiedApprovalPrincipal,
} from "../core/types.js";
import { ApprovalError } from "../core/types.js";

/** The frozen policy version, stamped on every decision. */
export const APPROVAL_POLICY_VERSION = "approval-policy/v0.4.0";

/** Reason code of an allowed decision (schema: authorization_reason_code='AUTHORIZED'). */
export const REASON_AUTHORIZED = "AUTHORIZED";

/** Accounting ladder rank: accountant < senior_accountant < controller. */
const LADDER_RANK: Record<AccountingRole, number> = {
	accountant: 1,
	senior_accountant: 2,
	controller: 3,
	// Tax roles sit OUTSIDE the ladder at rank 0 and can never dominate it.
	tax_reviewer: 0,
	authorized_tax_professional: 0,
	// v0.8.0 evidence-lifecycle roles also sit OUTSIDE the ladder at rank 0
	// (explicit-match only, exactly like tax roles; design §8.1).
	records_compliance_officer: 0,
	tenant_records_owner: 0,
	tax_responsible: 0,
	operational_accountant: 0,
};

const ASSURANCE_RANK: Record<AssuranceLevel, number> = {
	low: 0,
	standard: 1,
	strong: 2,
};

function denied(reasonCode: ApprovalErrorCode): ApprovalAuthorizationDecision {
	return {
		allowed: false,
		policyVersion: APPROVAL_POLICY_VERSION,
		reasonCode,
	};
}

/**
 * Base role by fiscal effect. Effects without a base role (none, or unknown)
 * return undefined — a memory without a fiscal effect cannot be approved.
 */
export function baseRoleForEffect(
	effect: FiscalEffect,
): AccountingRole | undefined {
	switch (effect) {
		case "journal_entry":
		case "adjustment":
		case "reclassification":
			return "accountant";
		case "approval":
			return "senior_accountant";
		case "closing":
			return "controller";
		case "declaration":
			return "tax_reviewer";
		case "sunat_filing":
			return "authorized_tax_professional";
		default:
			return undefined;
	}
}

/**
 * Raises the base role for the declared materiality level. Applies ONLY to
 * effects whose base role is accountant or senior_accountant; normal/NULL
 * stays at base; tax bases never raise.
 */
function requiredRole(
	base: AccountingRole,
	level: MaterialityLevel | undefined,
): AccountingRole {
	if (
		level === "critical" &&
		(base === "accountant" || base === "senior_accountant")
	) {
		return "controller";
	}
	if (level === "material" && base === "accountant") {
		return "senior_accountant";
	}
	return base;
}

/**
 * Role dominance exists ONLY within the accounting ladder; tax roles are
 * matched explicitly — a controller NEVER implies tax_reviewer or
 * authorized_tax_professional, and vice versa.
 */
function satisfies(
	roles: readonly AccountingRole[],
	required: AccountingRole,
): boolean {
	if (
		required === "tax_reviewer" ||
		required === "authorized_tax_professional"
	) {
		return roles.includes(required);
	}
	const rr = LADDER_RANK[required];
	return roles.some((role) => LADDER_RANK[role] >= rr);
}

/**
 * The pure v0.4.0 authorization decision for a pre-verified principal and a
 * memory. Returns the FIRST frozen reason code in the fixed check order.
 */
export function authorizeApproval(
	principal: VerifiedApprovalPrincipal,
	memory: AccountingMemory,
): ApprovalAuthorizationDecision {
	// 1. Tenant scope: the principal's tenant must equal the memory's tenant.
	if (memory.scope.kind === "institutional") {
		// An institutional scope carries no organizationId (Go: "").
		if (principal.tenantId !== "") {
			return denied("TENANT_SCOPE_MISMATCH");
		}
		return denied("COMPANY_SCOPE_DENIED");
	}
	if (principal.tenantId !== memory.scope.organizationId) {
		return denied("TENANT_SCOPE_MISMATCH");
	}
	// 2. Company scope: the memory's company must be inside the membership's
	//    company scopes.
	if (!principal.companyScopes.includes(memory.scope.companyId)) {
		return denied("COMPANY_SCOPE_DENIED");
	}
	// 3. Membership active: a pre-verified principal always carries the
	//    membership id of the ACTIVE membership it was derived from; empty
	//    means no active membership.
	if (principal.membershipId === "") {
		return denied("MEMBERSHIP_INACTIVE");
	}
	// 4. Role: the principal must satisfy the BASE role for the effect.
	const base = baseRoleForEffect(memory.fiscalEffect);
	if (base === undefined || !satisfies(principal.roles, base)) {
		return denied("ROLE_NOT_AUTHORIZED");
	}
	// 5. Assurance: minimum standard; sunat_filing additionally requires strong.
	const minAssurance: AssuranceLevel =
		memory.fiscalEffect === "sunat_filing" ? "strong" : "standard";
	if (ASSURANCE_RANK[principal.assuranceLevel] < ASSURANCE_RANK[minAssurance]) {
		return denied("ASSURANCE_TOO_LOW");
	}
	// 6. Materiality: the declared level may raise the requirement above the
	//    principal's reach → MATERIALITY_LIMIT_EXCEEDED.
	const raised = requiredRole(base, memory.materialityLevel);
	if (raised !== base && !satisfies(principal.roles, raised)) {
		return denied("MATERIALITY_LIMIT_EXCEEDED");
	}
	return {
		allowed: true,
		policyVersion: APPROVAL_POLICY_VERSION,
		reasonCode: REASON_AUTHORIZED,
	};
}

// ──────────────────────────────────────────────
// v0.9.0 review-workspace clauses (design §5/§6) — pure, fail-closed
// ──────────────────────────────────────────────

/**
 * Reports whether the authenticated reviewer IS the pending revision's
 * proposer — the separation-of-duties clause (design §5/§6.5.5): the decision
 * is only legal when the principal's subject id DIFFERS from the pending
 * revision's recordedBy. A non-empty proposer that equals the reviewer is a
 * violation (fail-closed inside the transaction); an empty proposer (a source
 * without an actor id) never equals an authenticated reviewer. Mirrors
 * authz.SODViolation.
 */
export function sodViolation(
	proposerRecordedBy: string,
	reviewerSubjectId: string,
): boolean {
	return proposerRecordedBy !== "" && proposerRecordedBy === reviewerSubjectId;
}

/**
 * Reports whether the declared materiality level demands the two review
 * checks (design §5/§6): material and critical require evidenceInspected &&
 * ruleInspected; normal/NULL never do. Mirrors authz.ReviewChecksRequired.
 */
export function reviewChecksRequired(
	level: MaterialityLevel | undefined,
): boolean {
	return level === "material" || level === "critical";
}

/**
 * Fails closed on the anti-rubber-stamp clause: when the declared materiality
 * level demands the two review checks, BOTH must be true
 * (REVIEW_CHECKS_REQUIRED otherwise). When no check is demanded, the checks
 * are ignored (a normal approval never trips this clause). Mirrors
 * authz.ValidateReviewChecks.
 */
export function validateReviewChecks(
	level: MaterialityLevel | undefined,
	checks: { evidenceInspected: boolean; ruleInspected: boolean },
): void {
	if (
		reviewChecksRequired(level) &&
		!(checks.evidenceInspected && checks.ruleInspected)
	) {
		throw new ApprovalError(
			"REVIEW_CHECKS_REQUIRED",
			"material/critical approvals require both review checks (evidenceInspected and ruleInspected)",
		);
	}
}
